### ISSUE TITLE
Register block template function fails in custom post type is prefixed

### ISSUE THREAD

**Issue by wpmark** [WordPress/gutenberg#67066]
### Description

When using the `register_block_template` function to register block templates, if the post type name you are registering a template for includes a prefix the registration doesn't work.

For example, the following declaration works:

```php
register_block_template( 'hd-people//single-person', [
	'title'       => __( 'Single Person', 'hd-people' ),
	'description' => __( 'Displays a single person profile.', 'hd-people' ),
	'content'     => hd_people_get_template_content( 'single-person.php' )
] );
```

However if the post type is prefixed with something, in this case `hd_` them the this fails. For example the following doesn't work.

```php
register_block_template( 'hd-people//single-hd_person', [
	'title'       => __( 'Single Person', 'hd-people' ),
	'description' => __( 'Displays a single person profile.', 'hd-people' ),
	'content'     => hd_people_get_template_content( 'single-hd_person.php' )
] );
```

Tagging @ndiego as requested.

### Step-by-step reproduction instructions

Create a custom post type with a prefix for its name/slug.

Register a block template using the `register_post_template()` function for the single post template.

The template never appears in the site editor under templates.

### Environment info

WordPress 6.7 with the Twenty Twenty Five Theme. No other plugins installed or active.

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [x] Yes

### Please confirm which theme type you used for testing.

- [x] Block
- [ ] Classic
- [ ] Hybrid (e.g. classic with theme.json)
- [ ] Not sure

**Comment by ndiego** [comment]
Thanks @wpmark. This issue is related to https://github.com/WordPress/gutenberg/pull/61577 and I have confirmed this bug. @Aljullu do you have any insights here?

**Comment by alexandrebuffet** [comment]
I've tried this new API with custom post types and I also use the underscore character “_” in their key so I get the same error.

This error seems to be displayed because one of statement for the template name check returns false in `register` method from `WP_Block_Templates_Registry` class:
```
elseif ( ! preg_match( '/^[a-z0-9-]+\/\/[a-z0-9-]+$/', $template_name ) ) {
    $error_message = __( 'Template names must contain a namespace prefix. Example: my-plugin//my-custom-template' );
    $error_code    = 'template_no_prefix';
} 
```

Updating the RegEx to allow the underscore character “_” should correct this problem?

**Comment by Aljullu** [comment]
Yes, we used the same RegExp that was used in the block registry. I don't have a strong opinion on this, but based on your comments it seems like a legit use case, so I agree on allowing the underscore character.

**Comment by alexandrebuffet** [comment]
Ok thanks @Aljullu, I can make the PR to correct this, should I create it here or on the core?

**Comment by karthick-murugan** [comment]
@wpmark , @alexandrebuffet  - I have add a patch file in the [Core Trac](https://core.trac.wordpress.org/ticket/62523#comment:1) with the fix and [testing video](https://share.cleanshot.com/qknmwCwGjX5tBWbCdFX5).

**Comment by mathetos** [comment]
I just stumbled on this bug today, glad to see it's already identified. One obvious place for which this is required in Core is if you are trying to register a block template for the post_tag taxonomy. 

If I have a custom taxonomy, I can register the block template fine like this:

```
register_block_template('cxdocs//taxonomy-sections', 
        [
            'title'       => __('Sections Archive Page Template', 'cxdocs'),
            'description' => __('A layout for CXDocs Section Archive.', 'cxdocs'),
            'content'     =>  $template_sections_content,
            'taxonomies'  => [ 'sections' ], // Target the Sections taxonomy
        ]
    );
```

But it doesn't work for Core's `post_tag` taxonomy at all:

```
register_block_template('cxdocs//taxonomy-post_tag', 
        [
            'title'       => __('Post Tag Archive Template', 'cxdocs'),
            'description' => __('A layout for CXDocs Section Archive.', 'cxdocs'),
            'content'     =>  $template_sections_content,
            'taxonomies'  => [ 'post_tag' ], // Target the post_tag taxonomy
        ]
    );
```

**Comment by karthick-murugan** [comment]
> I just stumbled on this bug today, glad to see it's already identified. One obvious place for which this is required in Core is if you are trying to register a block template for the post_tag taxonomy.
> 
> If I have a custom taxonomy, I can register the block template fine like this:
> 
> ```
> register_block_template('cxdocs//taxonomy-sections', 
>         [
>             'title'       => __('Sections Archive Page Template', 'cxdocs'),
>             'description' => __('A layout for CXDocs Section Archive.', 'cxdocs'),
>             'content'     =>  $template_sections_content,
>             'taxonomies'  => [ 'sections' ], // Target the Sections taxonomy
>         ]
>     );
> ```
> 
> But it doesn't work for Core's `post_tag` taxonomy at all:
> 
> ```
> register_block_template('cxdocs//taxonomy-post_tag', 
>         [
>             'title'       => __('Post Tag Archive Template', 'cxdocs'),
>             'description' => __('A layout for CXDocs Section Archive.', 'cxdocs'),
>             'content'     =>  $template_sections_content,
>             'taxonomies'  => [ 'post_tag' ], // Target the post_tag taxonomy
>         ]
>     );
> ```

@mathetos - I have checked the above code and the register_block_template works with or without the taxonomies args passed to it. So can you brief your issue little bit more so that I can get a clear idea of it?

**Comment by mathetos** [comment]
@karthick-murugan The bug I discovered isn't to do with the taxonomies argument, but with the template name parameter: i.e. `cxdocs//taxonomy-post_tag` . When that parameter is a single term like `taxonomy-posts` it works fine. But as soon as the term includes an underscore, the template doesn't get registered correctly. 

Here's a full example:

```
function cxdocs_register_post_tag_template() {

    $template_tag = plugin_dir_path(__FILE__) . 'templates/taxonomy-post_tag.html';
    $template_tag_content = file_exists($template_tag) ? file_get_contents($template_tag) : '';

    register_block_template('cxdocs//taxonomy-post_tag', 
        [
            'title'       => __('Post Tag Archive Template', 'cxdocs'),
            'description' => __('A layout for Post Tag Archive.', 'cxdocs'),
            'content'     =>  $template_tag_content,
        ]
    );

    wp_cache_flush();
}

add_action('init', 'cxdocs_register_post_tag_template', 1);
```

**Comment by karthick-murugan** [comment]
> [@karthick-murugan](https://github.com/karthick-murugan) The bug I discovered isn't to do with the taxonomies argument, but with the template name parameter: i.e. `cxdocs//taxonomy-post_tag` . When that parameter is a single term like `taxonomy-posts` it works fine. But as soon as the term includes an underscore, the template doesn't get registered correctly.

@mathetos - Thats the exact issue that we have fixed in the patch file mentioned in [this](https://core.trac.wordpress.org/ticket/62523) core ticket. Have you applied the patch file changes in your site and tested it? I have tested the code that you have provided and it is working fine. Please have a look at this video.

https://github.com/user-attachments/assets/a9dc1f22-ec74-46f0-b8a3-9ce1dc2f0e14

cc @carolinan
